### DISCUSSION TITLE
How to fine-tune the chatbot response

### DISCUSSION THREAD

**Discussion by Max-Falcon** [FlowiseAI/Flowise#1395] [Q&A] [answered]
I created a simple chat flow using Retrieval QA chain. First made an upserter chat, then called the vector source in another chatbot like the 2 images below. This works but I don't know what node i should use to add a simple fine tunning element to this chatbot. Basically I want to give some instructions to chat bot to return more refined answers. I found an example in the marketplace called "Conversational Retrieval Agent" which kind of does the trick but it looks like there is a limit in the length of instructions I can put in the Retriever Tool. Is there any other way to this? Thanks

![1-upserter](https://github.com/FlowiseAI/Flowise/assets/6592726/82858bf3-ba74-4ab7-96e6-6d0ab112bac9)
![2-chatbot](https://github.com/FlowiseAI/Flowise/assets/6592726/aa84e386-fe68-4faf-a2a0-7685d5264ff3)

**Comment by Jaredude** [comment]
Try using gpt-3.5-turbo-16k instead of just gpt-3.5. It's faster and allows for a much larger context. That will allow you to have a single, more complex prompt that performs what you're looking to do with plenty of room for retrieval context. Hope that helps.

**Reply by Max-Falcon** [reply, marked as answer]
Thanks for the reply. The 16k GPT is definitly helpful. But there was a lot more to it, at least in my project. I did a ton of research and try-n-error and finally was able to tame my bot. I used retrival agent instead of QA chain, efficiently prompted it, eductaed myself on how vector db and retrival works which led me to understand the importance of the actual document's text and content structure and how to fine tune the text. then I did a tone of testing using a chat analytic platform (langfuse, an awesome platform!), fine tuned the text and the prompt a bit more and now I get a decent percentage of correct answers with using very low tokents per question.

**Reply by Jaredude** [reply]
Nice! Sounds like a lot of hard work paid off!
